scripts/init.sh

```diff
@@ -28,6 +28,14 @@ HOST_IP=$(curl -H "X-aws-ec2-metadata-token: $TOKEN" http://169.254.169.254/late
 PRIMARY_IF=$(ip -4 -o a | grep "$HOST_IP" | awk '{print $2}')
 sysctl -w "net.ipv4.conf.$PRIMARY_IF.rp_filter=2"
 
+# Configure tcp_early_demux to enable kubelet to pod-eni communication
+# https://lwn.net/Articles/503420/ for background
+if [ "${ENABLE_TCP_EARLY_DEMUX:-false}" == "true" ]; then
+    sysctl -w "net.ipv4.tcp_early_demux=1"
+else
+    sysctl -w "net.ipv4.tcp_early_demux=0"
+fi
+
 cat "/proc/sys/net/ipv4/conf/$PRIMARY_IF/rp_filter"
```

Comment:
I think this `cat rp_filter` line should go above with the other `rp_filter` code, otherwise the output is going to be quite confusing :P

Reply:
I think we also want to add `env: [ {name: "ENABLE_TCP_EARLY_DEMUX", value: "false"} ]`, on line 229 of [config/master/manifests.jsonnet](https://github.com/aws/amazon-vpc-cni-k8s/blob/master/config/master/manifests.jsonnet#L229)

Reply:
ah true .. missed it, will update.

Reply:
> I think we also want to add env: [ {name: "ENABLE_TCP_EARLY_DEMUX", value: "false"} ], on line 229 of config/master/manifests.jsonnet

Re ^this:  Are we going to disable early_demux for everyone?  Only for sg-pp users?  Ask users to disable the option explicitly as part of the instructions for enabling sg-pp?

Reply:
We discussed about this, currently the plan is to do the performance testing between kubelet to other pods and then release it as part of 1.7.3. In couple of blog posts called out that actually disabling this would be beneficial (https://blog.packagecloud.io/eng/2016/06/22/monitoring-tuning-linux-networking-stack-receiving-data/#tuning-adjusting-ip-protocol-early-demux and http://www.newfreesoft.com/linux/linux_kernel_socket_protocol_stack_routing_lookup_cache_mechanism_1567/, which describes that caching happens at two places one through conntrack as well)

Making it disabled will also remove one more flag that needs to be turned on to enable sg-pp. We can also update the default value to true, if its causing performance issues on some of the clusters.

Reply:
Also adding to what @SaranBalaji90 commented - 

These are the function's in kernel which are executed ->

`skb_sk_is_empty -> 1 : sk is null and 0 : sk is not null`
```
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 |ip_rcv_finish -> 0 -> 0 [skb_sk_is_empty 1 ]  
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 |ip_route_input_rcu -> 0 -> 0 [skb_sk_is_empty 0 ] 
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 |e__ip_route_input_rcu:0.0.0.0/0 is FIB HIT [RTN_UNICAST(Gateway or direct route)] 
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 | e__ip_route_input_rcu -> 0 -> 0 [skb_sk_is_empty 0 ]
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 |ip_forward -> 0 -> 0 [skb_sk_is_empty 0 ] 
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 |e__ip_forward -> 1 -> 0 [skb_sk_is_empty 0 ] 
vlan990070403b2 | 192.168.78.231 -> 192.168.94.234 |e__ip_rcv_finish -> 1 -> 0 [skb_sk_is_empty 0 ] 
```

For the branch eni case it was going into both if conditions in `ip_rcv_finish`-
```
	if (net->ipv4.sysctl_ip_early_demux &&
	    !skb_dst(skb) &&
	    !skb->sk &&
	    !ip_is_fragment(iph)) {
		const struct net_protocol *ipprot;
		int protocol = iph->protocol;
		ipprot = rcu_dereference(inet_protos[protocol]);
		if (ipprot && (edemux = READ_ONCE(ipprot->early_demux))) {
			err = edemux(skb);
			if (unlikely(err))
				goto drop_error;
			/* must reload iph, skb->head might have changed */
			iph = ip_hdr(skb);
		}
	}
	/*
	 *	Initialise the virtual path cache for the packet. It describes
	 *	how the packet travels inside Linux networking.
	 */
	if (!skb_valid_dst(skb)) {
		err = ip_route_input_noref(skb, iph->daddr, iph->saddr,
					   iph->tos, dev);
		if (unlikely(err))
			goto drop_error;
	}
```
Because the first if does - `!skb_dst(skb) && !skb->sk` so that means `sk is null` and `dst is null`, so now early demux filled dst_entry. But then `skb_valid_dst` is failing so now it also does a route lookup which in turn calls `ip_forward` and because `skb->sk` is set then packet is dropped. Basically the early demux optimization purpose of using the cached value rather than routing lookup wasn't working here.